### PR TITLE
[GPII-3325]: Use application-default credentials 

### PR DESCRIPTION
**Pull request by natarajaya**
This is the implementation of alternative to https://github.com/gpii-ops/gpii-infra/pull/225.
I noticed only couple unwanted side-effects:
- Looks like there is no way to make `gcloud` CLI to use application-default credentials, which means that we still need to login normally using `gcloud auth login`. Not a big deal, but a slight inconvenience.
- When application-default credential are being used for `google-cloud-monitoring` gem authentication, I am receiving this warning:
```
null_resource.apply_stackdriver_monitoring (local-exec): Your application has authenticated using end user credentials from Google Cloud SDK. We recommend that most server applications use service accounts instead. If your application continues to use end user credentials from Cloud SDK, you might receive a "quota exceeded" or "API not enabled" error. For more information about service accounts, see https://cloud.google.com/docs/authentication/.
```
I never received mentioned "quota exceeded" or "API not enabled" errors, despite running `rake` quite often, so not sure if this is going to cause any issues for developers.

### CODE REVIEW

**Comment by natarajaya** [comment]
> So does this actually need dual login during normal operations?

@stepanstipl Yes. There are bunch of places where `gcloud` being invoked directly from code, replacing them all with API calls does not seem feasible.

> I remember you mentioned that it'll be necessary to change some print-access-token call, I don't see that change here, is it not required anymore?

That call is from Secrets module, since we can not get away from the need to login normally, this change is not needed.

**Comment by stepanstipl** [comment]
Thanks @natarajaya

> That call is from Secrets module, since we can not get away from the need to login normally, this change is not required.

How does it matter where the call is from? It should not make a difference, or?

**Comment by natarajaya** [comment]
> How does it matter where the call is from? It should not make a difference, or?

@stepanstipl It does not matter. We need to login, because we have `gcloud` calls in many places, and since we need to login, we can keep using `print-access-token` the same way.

File: shared/rakefiles/xk_config.rake [hunk 1]
```diff
@@ -15,19 +16,30 @@ end
...
-  "
+  unless File.file?(@serviceaccount_key_file)
+    sh "
+      gcloud iam service-accounts keys create #{@serviceaccount_key_file} \
+        --iam-account projectowner@$TF_VAR_project_id.iam.gserviceaccount.com
+    "
+  end
+end
+
+@default_creds_file = "/root/.config/gcloud/application_default_credentials.json"
```
Comment (mrtyler) [inline shared/rakefiles/xk_config.rake]: I find this name confusing. How about `@application_default_credentials_file`, `@application_default_creds_file`, `@app_default_creds_file`, or `@adc_file`?

File: shared/rakefiles/xk_config.rake [hunk 2]
```diff
@@ -15,19 +16,30 @@ end
...
+task :configure_app_default_login => [@default_creds_file]
+rule @default_creds_file do
+  # This retrieves application-default credentials using interactive auth,
+  # only in case service account credentials are not present.
+  unless File.file?(@serviceaccount_key_file)
```
Comment (mrtyler) [inline shared/rakefiles/xk_config.rake]: What do you think of creating an empty `@default_creds_file` (or, perhaps better, a file containing a comment like `# Empty application default credentials file because there is a serviceaccount key file instead`) so this task doesn't run every time?

Or maybe some kind of message when we avoid creating an ADC because we already have an SA key? I feel like it could be easy to get mixed up and end up with weird behavior.

Maybe I'm worrying too much and/or prematurely?

File: shared/rakefiles/xk_util.rake [hunk 3]
```diff
@@ -51,3 +51,22 @@ task :rotate_secrets_key, [:encryption_key] => [:configure_serviceaccount] do |t
   Secrets.set_secrets(@secrets, rotate_secrets = true)
   Secrets.disable_non_primary_key_versions(args[:encryption_key], new_version_id)
 end
+
+# This task destroy all keys except current one for projectowner's SA.
+# All operations will be ignored in case local credentials not present or SA usage not configured.
+task :destroy_sa_keys => [:configure] do
+  sh "
+    if [ \"$TF_VAR_serviceaccount_key\" != \"\" ] && [ -f $TF_VAR_serviceaccount_key ]; then \
```
Comment (mrtyler) [inline shared/rakefiles/xk_util.rake]: Not sure if it matters but `-f` behaves differently from Ruby `File.file?`

>        -f file
>               True if file exists and is a regular file.

File: shared/rakefiles/xk_util.rake [hunk 4]
```diff
@@ -51,3 +51,22 @@ task :rotate_secrets_key, [:encryption_key] => [:configure_serviceaccount] do |t
   Secrets.set_secrets(@secrets, rotate_secrets = true)
   Secrets.disable_non_primary_key_versions(args[:encryption_key], new_version_id)
 end
+
+# This task destroy all keys except current one for projectowner's SA.
+# All operations will be ignored in case local credentials not present or SA usage not configured.
```
Comment (mrtyler) [inline shared/rakefiles/xk_util.rake]: How about this wording: "This task does nothing unless local credentials are present or a Serviceaccount is configured."

File: shared/rakefiles/xk_config.rake [hunk 5]
```diff
@@ -38,12 +50,15 @@ rule @kubectl_creds_file => [@gcp_creds_file] do
...
-    existing_cluster=$(gcloud container clusters list --filter #{cluster_name} --zone #{zone} --project #{ENV["TF_VAR_project_id"]})
+    existing_cluster=$(gcloud container clusters list --filter #{cluster_name} --zone #{zone} --project $TF_VAR_project_id)
     if [ $? == 0 ] && [ \"$existing_cluster\" != \"\" ]; then \
-      gcloud container clusters get-credentials #{cluster_name} --zone #{zone} --project #{ENV["TF_VAR_project_id"]}
+      gcloud container clusters get-credentials #{cluster_name} --zone #{zone} --project $TF_VAR_project_id
     fi"
 end
 
-task :configure_current_project => [@gcp_creds_file] do
+task :configure => [@gcp_creds_file, @default_creds_file, @kubectl_creds_file] do
```
Comment (mrtyler) [inline shared/rakefiles/xk_config.rake]: How about `:configure_creds`?

File: shared/rakefiles/xk_config.rake [hunk 6]
```diff
@@ -38,12 +50,15 @@ rule @kubectl_creds_file => [@gcp_creds_file] do
...
   sh "gcloud config set project $TF_VAR_project_id"
+  # We need to set service account key var only if SA credentials are present locally.
+  # In case var is unset, application-default credentials will be used.
+  ENV['TF_VAR_serviceaccount_key'] = File.file?(@serviceaccount_key_file) ? @serviceaccount_key_file : ''
```
Comment (mrtyler) [inline shared/rakefiles/xk_config.rake]: Please change all these single quotes to double quotes.

File: shared/rakefiles/xk_infra.rake [hunk 7]
```diff
@@ -38,7 +38,7 @@ task :refresh_common_infra, [:project_type] => [@gcp_creds_file] do | taskname,
   end
 end
 
-task :apply_common_infra => [@gcp_creds_file] do
+task :apply_common_infra => [:configure] do
   # Steps to initialize GCP with a minimum set of resources to allow Terraform
   # create the rest of the infrastructure.
   # These steps are the same found in this tutorial:
```
Comment (mrtyler) [inline shared/rakefiles/xk_infra.rake]: Looks line Line 60 (to which Github won't allow me to attach a comment) still refers to `:configure_current_project`.

File: shared/rakefiles/entrypoint.rake [hunk 8]
```diff
@@ -68,22 +66,15 @@ task :set_compose_env do
...
 end
 
-desc "[ADVANCED] Authenticate and generate GCP credentials (gcloud auth login)"
-task :configure_login => [:set_vars] do
-  sh "#{@exekube_cmd} rake configure_login"
-end
-
-# This duplicates information in docker-compose.yaml,
-# TF_VAR_serviceaccount_key.
 desc "[ADVANCED] Create and download credentials for projectowner service account"
```
Comment (mrtyler) [inline shared/rakefiles/entrypoint.rake]: Maybe this should say `[CI ONLY]` now?

File: shared/rakefiles/xk_config.rake (hunk 1 above)
Comment (stepanstipl) [reply shared/rakefiles/xk_config.rake]: +1, `@application_default_credentials_file`

File: shared/rakefiles/xk_config.rake (hunk 2 above)
Comment (stepanstipl) [reply shared/rakefiles/xk_config.rake]: IMHO file present/not present is sufficient mechanism here (involving actual content in the check just makes it unnecessary complicated)

**Comment by amatas** [comment]
I've review the comments and the code, I have nothing to add, so LGTM once all the comments from Stepan and Tyler are fixed.

File: shared/rakefiles/xk_config.rake [hunk 9]
```diff
@@ -15,19 +16,30 @@ end
...
   # gcp-project-init). If gcp-project-init becomes idempotent (GPII-2989,
   # https://github.com/exekube/exekube/issues/92), or if this 'keys create'
   # step moves somewhere else in exekube, call this command from that place
   # instead.
-  sh "
-    [ -f $TF_VAR_serviceaccount_key ] || \
-    gcloud iam service-accounts keys create $TF_VAR_serviceaccount_key \
-      --iam-account projectowner@$TF_VAR_project_id.iam.gserviceaccount.com
-  "
+  unless File.file?(@serviceaccount_key_file)
```
[the remark this answers is not in the document]
Comment (natarajaya) [reply shared/rakefiles/xk_config.rake]: Ruby is great!

File: shared/rakefiles/xk_config.rake [hunk 10]
```diff
@@ -1,3 +1,4 @@
+
```
[the remark this answers is not in the document]
Comment (natarajaya) [reply shared/rakefiles/xk_config.rake]: Removed.

File: shared/rakefiles/xk_config.rake (hunk 1 above)
Comment (natarajaya) [reply shared/rakefiles/xk_config.rake]: I choose `@app_default_creds_file` to be match style of `@gcp_creds_file`.

File: shared/rakefiles/xk_config.rake (hunk 2 above)
Comment (natarajaya) [reply shared/rakefiles/xk_config.rake]: I am with @stepanstipl here.
To avoid confusion which auth method is being used, I'll add some debug output.

File: shared/rakefiles/xk_util.rake (hunk 3 above)
Comment (natarajaya) [reply shared/rakefiles/xk_util.rake]: It behaves the same way, but `File.file?` allows to make output cleaner by not spitting SA configuration commands on every run.

File: shared/rakefiles/xk_util.rake (hunk 4 above)
Comment (natarajaya) [reply shared/rakefiles/xk_util.rake]: Changed to `It does nothing in case local SA credentials not present.`

File: shared/rakefiles/xk_config.rake (hunk 5 above)
Comment (natarajaya) [reply shared/rakefiles/xk_config.rake]: I would like to keep it as `:configure` and add more stuff (like [auth user email setting](https://github.com/gpii-ops/gpii-infra/pull/225/files#diff-9bf20f39183745233b6ee4d09596753fR70)) that requires to be executed on every run.

File: shared/rakefiles/xk_config.rake (hunk 6 above)
Comment (natarajaya) [reply shared/rakefiles/xk_config.rake]: Done.

File: shared/rakefiles/entrypoint.rake (hunk 8 above)
Comment (natarajaya) [reply shared/rakefiles/entrypoint.rake]: Agree. Updated comment and moved to `ci.rake`.

File: shared/rakefiles/xk_infra.rake (hunk 7 above)
Comment (natarajaya) [reply shared/rakefiles/xk_infra.rake]: Nice catch! Fixed.

**Comment by natarajaya** [comment]
> Any changes required in ci.rake?

@mrtyler I moved `:configure_serviceaccount` task there, no changes required beside this.

> Something that I think will help us understand the workflow is some high-level docs.

Let me add authentication workflow section into GCP readme.

**Comment by natarajaya** [comment]
@mrtyler [Updated README](https://github.com/gpii-ops/gpii-infra/blob/32295e8cc8f1a1174ee975365477bc4b0393c627/gcp/README.md#authentication-workflow).

File: shared/rakefiles/xk_util.rake (hunk 3 above)
Comment (mrtyler) [reply shared/rakefiles/xk_util.rake]: It does not behave the same way with respect to links. I mentioned it in the context of [this other comment](https://github.com/gpii-ops/gpii-infra/pull/229#discussion_r241257190).

File: shared/rakefiles/xk_config.rake [hunk 11]
```diff
@@ -32,31 +31,38 @@ task :configure_serviceaccount => [@gcp_creds_file] do
...
+task :configure => [@gcp_creds_file, @app_default_creds_file, @kubectl_creds_file] do
   sh "gcloud config set project $TF_VAR_project_id"
   # We need to set service account key var only if SA credentials are present locally.
   # In case var is unset, application-default credentials will be used.
-  ENV['TF_VAR_serviceaccount_key'] = File.file?(@serviceaccount_key_file) ? @serviceaccount_key_file : ''
+  ENV["TF_VAR_serviceaccount_key"] = File.file?(@serviceaccount_key_file) ? @serviceaccount_key_file : ""
+  # Setting authenticated user's email into env variable, so it can be
+  # accessible in modules: https://issues.gpii.net/browse/GPII-3516
+  ENV["TF_VAR_auth_user_email"] = %x{
+    gcloud auth list --filter='account!~gserviceaccount.com' --format json |  jq -r '.[].account'
```
Comment (mrtyler) [inline shared/rakefiles/xk_config.rake]: I guess I asked about this on the other PR. Can you either limit this to the first result, or return all results as a terraform list? I prefer being explicit to unexpected results.

**Comment by stepanstipl** [comment]
LGTM

File: shared/rakefiles/xk_config.rake (hunk 11 above)
Comment (natarajaya) [reply shared/rakefiles/xk_config.rake]: Done.
